### PR TITLE
Fix the tests, and the apiserver behaviour

### PR DESCRIPTION
**Pull request by howbazaar**
Comments left on PR to outline reasons.

### CODE REVIEW

File: apiserver/export_test.go [hunk 1]
```diff
@@ -26,7 +26,6 @@ var (
 	BZMimeType            = bzMimeType
 	JSMimeType            = jsMimeType
 	GUIURLPathPrefix      = guiURLPathPrefix
-	SpritePath            = spritePath
```
Comment (howbazaar) [inline apiserver/export_test.go:29]: The spritePath was removed in the main source file, so needed to be removed from the tests as well

File: apiserver/gui.go [hunk 2]
```diff
@@ -49,15 +49,11 @@ var (
 //   "jujugui" directory where the actual Juju GUI files live;
 // - the "jujugui" directory includes a "static" subdirectory with the Juju
 //   GUI assets to be served statically;
-// - the "jujugui" directory includes a "templates/index.html.go" file which is
```
Comment (howbazaar) [inline apiserver/gui.go:52]: Updated the comments here to match implementation.

File: apiserver/gui.go [hunk 3]
```diff
@@ -86,8 +82,8 @@ func guiEndpoints(pattern, dataDir string, ctxt httpContext) []apihttp.Endpoint
 			})
 		}
 	}
-	add("/config.js", (*guiHandler).serveConfig)
-	add("/static/", (*guiHandler).serveStatic)
+	add(pattern+"config.js", (*guiHandler).serveConfig)
```
Comment (howbazaar) [inline apiserver/gui.go:85]: The registration of the patterns was missing the /dashboard/ prefix.

File: apiserver/gui.go [hunk 4]
```diff
@@ -140,9 +136,9 @@ func (gr *guiRouter) ensureFiles(req *http.Request) (rootDir string, hash string
 	if err != nil {
 		return "", "", errors.Trace(err)
 	}
-	logger.Debugf("serving Juju GUI version %s", vers)
+	logger.Tracef("serving Juju Dashboard version %s", vers)
```
Comment (howbazaar) [inline apiserver/gui.go:139]: Changed the Debug to Trace as we don't want to see this all the time in the logs. Also renamed GUI to dashboard in many places.

File: apiserver/gui.go [hunk 5]
```diff
@@ -245,10 +243,11 @@ type guiHandler struct {
 
 // serveStatic serves the GUI static files.
 func (h *guiHandler) serveStatic(w http.ResponseWriter, req *http.Request) {
-	logger.Debugf("serving Juju GUI static files")
 	staticDir := filepath.Join(h.rootDir, "static")
+	logger.Tracef("serving Juju Dashboard static files from %q", staticDir)
 	fs := http.FileServer(http.Dir(staticDir))
-	http.StripPrefix("/static/", fs).ServeHTTP(w, req)
+	prefix := path.Join(h.basePath, "static")
```
Comment (howbazaar) [inline apiserver/gui.go:249]: Again missing the /dashboard prefix in the prefix to strip

File: apiserver/gui.go [hunk 6]
```diff
@@ -286,13 +285,17 @@ func sendGUIComboFile(w io.Writer, fpath string) {
 
 // serveIndex serves the GUI index file.
 func (h *guiHandler) serveIndex(w http.ResponseWriter, req *http.Request) {
-	logger.Debugf("serving Juju Dashboard index")
+	logger.Tracef("serving Juju Dashboard index")
 	indexFile := filepath.Join(h.rootDir, "index.html")
-	t, err := template.ParseFiles(indexFile)
+
+	b, err := ioutil.ReadFile(indexFile)
```
Comment (howbazaar) [inline apiserver/gui.go:291]: Not a template, so don't use the template methods, just read the file and write out. Could potentially use io.Copy, but didn't.

File: apiserver/gui.go [hunk 7]
```diff
@@ -286,13 +285,17 @@ func sendGUIComboFile(w io.Writer, fpath string) {
...
 	if err != nil {
-		writeError(w, errors.Annotate(err, "cannot parse index file"))
+		writeError(w, errors.Annotate(err, "cannot read index file"))
+		return
+	}
+	if _, err := w.Write(b); err != nil {
+		writeError(w, errors.Annotate(err, "cannot write index file"))
 	}
-	writeError(w, errors.Annotate(t.Execute(w, nil), "cannot render index file"))
```
Comment (howbazaar) [inline apiserver/gui.go:295]: Don't write a nil error if there is no error.

File: apiserver/gui.go [hunk 8]
```diff
@@ -406,12 +409,6 @@ func writeError(w http.ResponseWriter, err error) {
 	}
 }
 
-// hashedPath returns the full path (including the GUI archive hash) to the
-// given path, that must not start with a slash.
-func (h *guiHandler) hashedPath(p string) string {
```
Comment (howbazaar) [inline apiserver/gui.go:411]: Remove the now unused function

File: apiserver/gui_test.go [hunk 9]
```diff
@@ -22,20 +22,18 @@ import (
...
 	apitesting "github.com/juju/juju/apiserver/testing"
 	"github.com/juju/juju/state/binarystorage"
 	"github.com/juju/juju/testing/factory"
-	jujuversion "github.com/juju/juju/version"
 )
 
 const (
-	guiConfigPath = "templates/config.js.go"
-	guiIndexPath  = "templates/index.html.go"
+	guiConfigPath = "config.js.go"
```
Comment (howbazaar) [inline apiserver/gui_test.go:35]: Not in template dir any more.

File: apiserver/gui_test.go [hunk 10]
```diff
@@ -22,20 +22,18 @@ import (
...
+	guiIndexPath  = "index.html"
```
Comment (howbazaar) [inline apiserver/gui_test.go:36]: Remove the ".go" bit too

File: apiserver/gui_test.go [hunk 11]
```diff
@@ -47,7 +45,7 @@ var _ = gc.Suite(&guiSuite{})
 // guiURL returns the complete URL where the Juju GUI can be found, including
 // the given hash and pathAndquery.
 func (s *guiSuite) guiURL(hash, pathAndquery string) string {
-	return s.urlFromBase(apiserver.GUIURLPathPrefix, hash, pathAndquery)
+	return s.urlFromBase(apiserver.GUIURLPathPrefix, "", pathAndquery)
```
Comment (howbazaar) [inline apiserver/gui_test.go:48]: No hash any more in the path.

File: apiserver/gui_test.go [hunk 12]
```diff
@@ -154,51 +152,16 @@ var guiHandlerTests = []struct {
...
-	setup: func(c *gc.C, baseDir string, storage binarystorage.Storage) string {
-		setupGUIArchive(c, storage, "2.0.42", map[string]string{
-			apiserver.SpritePath: "",
-		})
-		return ""
-	},
-	currentVersion: "2.0.42",
-	expectedStatus: http.StatusInternalServerError,
-	expectedError:  "cannot parse template: .*: no such file or directory",
-}, {
```
Comment (howbazaar) [inline apiserver/gui_test.go:178]: Don't care about invalid templates for index as it isn't a template any more.

File: apiserver/gui_test.go [hunk 13]
```diff
@@ -219,81 +182,6 @@ var guiHandlerTests = []struct {
...
-	pathAndquery:   "combo?foo&%%",
-	expectedStatus: http.StatusBadRequest,
-	expectedError:  `cannot combine files: invalid file name "%": invalid URL escape "%%"`,
-}, {
-	about: "combo: invalid file path",
-	setup: func(c *gc.C, baseDir string, storage binarystorage.Storage) string {
-		return setupGUIArchive(c, storage, "1.0.0", nil)
-	},
-	currentVersion: "1.0.0",
-	pathAndquery:   "combo?../../../../../../etc/passwd",
```
Comment (howbazaar) [inline apiserver/gui_test.go:247]: no more combos either.

File: apiserver/gui_test.go [hunk 14]
```diff
@@ -583,18 +413,12 @@ func (s *guiSuite) TestGUIConfig(c *gc.C) {
...
-    base: '{{.base}}',
-    bakeryEnabled: {{.bakeryEnabled}},
-    host: '{{.host}}',
-    controllerSocket: '{{.controllerSocket}}',
-    charmstoreURL: '{{.charmstoreURL}}',
-    socket: '{{.socket}}',
-    staticURL: '{{.staticURL}}',
-    uuid: '{{.uuid}}',
-    version: '{{.version}}'
+    baseAppURL: '{{.baseAppURL}}',
```
Comment (howbazaar) [inline apiserver/gui_test.go:420]: Match the template args with what we use in the apiserver code

File: cmd/juju/commands/main_test.go [hunk 15]
```diff
@@ -129,7 +129,7 @@ func (s *MainSuite) TestRunMain(c *gc.C) {
 		summary: "unknown command with match",
 		args:    []string{"discombobulate"},
 		code:    1,
-		out:     missingCommandMessage("discombobulate", "diff-bundle"),
+		out:     missingCommandMessage("discombobulate", "dashboard"),
```
Comment (howbazaar) [inline cmd/juju/commands/main_test.go:132]: The Levenshtein distance is now closer to dashboard.

File: cmd/juju/commands/main_test.go [hunk 16]
```diff
@@ -502,6 +502,7 @@ var commandNames = []string{
 	"create-storage-pool",
 	"create-wallet",
 	"credentials",
+	"dashboard",
```
Comment (howbazaar) [inline cmd/juju/commands/main_test.go:505]: Needed to add the new commands.

File: cmd/juju/gui/gui_test.go [hunk 17]
```diff
@@ -103,7 +103,7 @@ func (s *guiSuite) TestGUISuccessWithCredential(c *gc.C) {
 	out, err := s.run(c)
 	c.Assert(err, jc.ErrorIsNil)
 	c.Assert(out, jc.Contains, `
-Your login credential is:
+Your login credentials are:
```
Comment (howbazaar) [inline cmd/juju/gui/gui_test.go:106]: Need to update the tests when you change wording elsewhere.

File: environs/gui/simplestreams.go [hunk 18]
```diff
@@ -149,8 +149,8 @@ func appendArchives(
 		if err != nil {
 			return nil, errors.Annotate(err, "cannot parse supported juju version")
 		}
-		if majorVersion == supportedVersion.Major &&
-			minorVersion == supportedVersion.Minor {
+		if majorVersion != supportedVersion.Major ||
```
Comment (howbazaar) [inline environs/gui/simplestreams.go:152]: The logic here was the wrong way around.

File: environs/gui/simplestreams_test.go [hunk 19]
```diff
@@ -63,84 +63,51 @@ var fetchMetadataTests = []struct {
...
-		Size:    474747,
-		SHA256:  "5236f1b694a9a66dc4f86b740371408bf4ddf2354ebc6e5410587843a1e55743",
+		Version:              version.MustParse("2.4.0"),
+		SupportedJujuVersion: "2.8.0",
+		Path:                 "gui/2.4.0/jujugui-2.4.0.tar.bz2",
+		Size:                 6098111,
+		SHA256:               "6cec58b36969590d3ff56279a2c63b4f5faf277b0dbeefe1106f666582575894",
+	}},
+}, {
+	about:       "devel version 2.9.0",
```
Comment (howbazaar) [inline environs/gui/simplestreams_test.go:102]: Test against a later version

File: environs/gui/simplestreams_test.go [hunk 20]
```diff
@@ -63,84 +63,51 @@ var fetchMetadataTests = []struct {
...
 	// The following fields are automatically pre-populated by the test:
-	// "FullPath", "Source", "StringVersion" and "JujuMajorVersion"
+	// "FullPath", "Source", "DashboardVersion" and "SupportedJujuVersion"
 	expectedMetadata []*gui.Metadata
 	// expectedError optionally holds the expected error returned while trying
 	// to retrieve GUI metadata information.
 	expectedError string
 }{{
-	about:       "released version 2",
+	about:       "released version 2.8.2",
```
Comment (howbazaar) [inline environs/gui/simplestreams_test.go:72]: Test against a point release that is different

File: apiserver/gui.go (hunk 3 above)
Comment (hatched) [reply apiserver/gui.go:85]: Unfortunately the dashboard application requires that the config and static paths be served from the host directly. This worked before so I'll have to see if I can revert this change and update the tests.

File: environs/gui/simplestreams.go (hunk 18 above)
Comment (hatched) [reply environs/gui/simplestreams.go:152]: Woops, thanks :)

**Review by hatched — APPROVED** [review]
Thanks a lot for this! I'll see what I can do wrt the basepath once this lands.

**Pull request merged**